src/ui/3d/grid.js

```diff
@@ -7,17 +7,178 @@
+    rows: PropTypes.number.isRequired,
+    cols: PropTypes.number.isRequired,
+    cellSize: PropTypes.number,
+    thickness: PropTypes.number,
+    padding: PropTypes.number,
+    colorMap: PropTypes.object,
+    onClick: PropTypes.func,
+    onMouseOver: PropTypes.func,
+    onMouseOut: PropTypes.func,
+    children: PropTypes.oneOfType([
+      PropTypes.arrayOf(PropTypes.element),
+      PropTypes.element,
+    ]),
+  };
+  static defaultProps = {
+    colorMap: {},
+    cellSize: 1,
+    padding: 0.1,
+    thickness: 0.1,
+  };
+
+  constructor(props) {
+    super(props);
+    this.boardGroup = new THREE.Group();
+    this.tokenGroup = new THREE.Group();
+    this.boardGroup.add(this.tokenGroup);
+    // translate the board to center on (0,0,0)
+    this.boardGroup.translateX(
+      (-(this.props.padding + this.props.cellSize) * (this.props.cols - 1)) / 2
+    );
+    this.boardGroup.translateZ(
+      (-(this.props.padding + this.props.cellSize) * (this.props.rows - 1)) / 2
+    );
+  }
+
+  _getCellColor(x, y) {
+    const key = `${x},${y}`;
+    let color = '#777777';
+    if (key in this.props.colorMap) {
+      color = this.props.colorMap[key];
+    }
+    return color;
+  }
+
+  _drawSquare(ctx) {
+    this.ctx = ctx;
+    ctx.add(this.boardGroup);
+
+    // when rerendering, render a new squareGroup
+    this.boardGroup.remove(this.squareGroup);
+    this.squareGroup = new THREE.Group();
+    this.boardGroup.add(this.squareGroup);
+
+    // add square base
+    for (let x = 0; x < this.props.cols; x++) {
+      for (let y = 0; y < this.props.rows; y++) {
+        let squareProps = {
+          x: x,
+          y: y,
+          ctx: ctx,
+          size: this.props.cellSize,
+          color: this._getCellColor(x, y),
+          padding: this.props.padding,
+          thickness: this.props.thickness,
+        };
+        let square = new Square(squareProps);
+        this.squareGroup.add(square);
+        const onEvent = e => {
+          if (e.type == 'click') {
+            if (this.props.onClick) this.props.onClick({ x: x, y: y });
+          } else if (e.type == 'mouseOver') {
+            if (this.props.onMouseOver) this.props.onMouseOver({ x: x, y: y });
+          } else if (e.type == 'mouseOut') {
+            if (this.props.onMouseOut) this.props.onMouseOut({ x: x, y: y });
+          }
+        };
+        ctx.regCall(square, onEvent);
+      }
+    }
+    //set tokens
+    const tokens = React.Children.map(this.props.children, child => {
+      return React.cloneElement(child, {
+        three: true,
+        boardSize: this.props.cellSize,
+        ui: ctx,
+        parrent: this.tokenGroup,
+        padding: this.props.padding,
+        lift: this.props.thickness,
+      });
+    });
+
+    return tokens;
+  }
+
+  componentWillUnmount() {
+    this.ctx.remove(this.boardGroup);
+  }
+
   render() {
-    return null;
+    return (
+      <UIContext.Consumer>{ctx => this._drawSquare(ctx)}</UIContext.Consumer>
```

Comment:
Let's call this `context` to avoid confusing it with boardgame.io's `ctx`? Also take a look at deck.js which splits the context component into a separate wrapper. Let's use the same style here for consistency?